### PR TITLE
perf(core): Automated performance tuning by Claude

### PR DESCRIPTION
**Pull request by yamadashy**
## Summary

Automated performance tuning of the repomix pack pipeline. This PR consolidates multiple optimizations achieving cumulative improvements of **~25% on end-to-end CLI execution**.

### Previous commits (already merged into this branch)
1. **Parallel git log for file sorting** — overlap `git log` with file collection (~20% faster)
2. **Fast base64 detection** — linear character scan pre-check before regex (~5% faster)
3. **Adaptive IPC batch sizing** — reduce metrics worker overhead (~6% faster)
4. **Early metrics worker pool init** — overlap gpt-tokenizer loading with I/O phases (~4% faster)
5. **Synchronous file reads** — replace async readFile with readFileSync in collectFiles (~7% faster)
6. **Non-blocking cleanup + lazy computation** — fire-and-forget worker teardown, skip unnecessary work (~8% theoretical)
7. **Token estimation from file tokens** — estimate output tokens from file token counts + overhead ratio (~20% faster)
8. **git ls-files for file search** — replace globby&#39;s gitignore traversal with git index access (~7% faster)
9. **Lazy-load globby and avoid globby directory scan on git fast path** — replace the globby directory scan with lightweight directory discovery from git ls-files output when searching for empty directories. Lazy-load globby so it&#39;s never imported on the common git fast path.
10. **Two-phase base64 truncation pre-check** — adds `hasLongLine` (indexOf-based) + `hasLongNonWhitespaceRun` gating to skip expensive regex scans. Reduces truncateBase64Content from ~82ms to ~12ms for 1000 files.
11. **Fix: Move try block to cover metrics worker pool lifecycle** — ensures cleanup happens even if searchFiles or sortPaths throws.
12. **Skip redundant output tokenization** — when all file tokens are available, skip redundant output token counting.
13. **Reduce metrics pipeline latency** — batch size tuning and deferred output pattern for overlapping metrics with output generation.
14. **Security check keyword pre-filter** — Skip expensive `lintSource()` for files without secret-related keywords.
15. **Pre-create metrics worker pool before pack()** — Move metrics runner creation to defaultAction for earlier BPE warmup.
16. **Main-thread security check + pipeline restructuring** — Avoid spawning security worker pool when few items pass the keyword pre-filter (≤50 items, the common case). Security worker threads compete for CPU with metrics warmup workers on machines with ≤4 cores.
17. **Skip tokenizing all files when tokenCountTree is disabled** — Only tokenize the top-N largest files by character count to compute the token/char ratio and top-N token ranking.
18. **Replace sequential keyword scanning with single-pass regex** — The `mightContainSecret` pre-filter was using sequential `String.includes()` calls for ~50 keywords, scanning each file&#39;s content 50+ times. Replaced with a single pre-compiled `RegExp` alternation that scans content in one pass (3.5x speedup).
19. **Extract `mightContainSecret` into `securityPreFilter.ts` module** — Moved `mightContainSecret()` and related types out of `securityCheckWorker.ts` into a new `securityPreFilter.ts` module with zero dependencies on secretlint.
20. **Replace sequential keyword scanning with single-pass regex** — Combined regex for 3.5x faster pre-filter.
21. **Pre-load BPE data on main thread to eliminate redundant per-worker file I/O** — Each of the 4 metrics worker threads independently loaded gpt-tokenizer&#39;s BPE rank data (~3.6MB, 200K entries) from disk, taking 210-330ms per worker. Now the BPE data is loaded once on the main thread.
22. **Add unit tests for prefetchFileChangeCounts cache behavior**
23. **Combine picomatch patterns into single regex and lazy-load minimatch**
24. **Line-length pre-filter for base64 scan** — 85% faster base64 detection.
25. **Skip calculateFileLineCounts for non-skill output paths**
26. **Lazy-load handlebars, fast-xml-builder, and @clack/prompts**
27. **Remove pre-filter keywords for disabled secretlint rules**
28. **Replace O(N²) split output algorithm with O(N) estimation-based approach**
29. **Defer metrics worker pool creation until after file search**
30. **Run pack() inline instead of spawning child_process worker** — Eliminated ~250ms overhead.
31. **Size metrics worker pool by estimated tokenization needs** — Pool sized by actual work (1 thread for default 50-file case), eliminating unnecessary BPE init contention.
32. **Reduce startup module loading by deferring heavy imports** — Extract `TOKEN_ENCODINGS` to `tokenEncodings.ts`, lazy-load `json5`, preload `@secretlint/core` before metrics workers.

### Latest commit: Preload BPE data during searchFiles to eliminate metrics warmup stall

33. **Preload BPE data during searchFiles to eliminate metrics warmup stall** — The `loadBpeRanks` async I/O (~100ms) was called inside `createMetricsTaskRunner` (after searchFiles), but its callback couldn't fire during the subsequent synchronous `collectFiles` (~150ms), creating a ~200ms pipeline stall at `await metricsWarmupPromise`. Fix: fire `loadBpeRanks()` at the start of `pack()` so it overlaps with searchFiles I/O, pass the pre-loaded promise to `createMetricsTaskRunner`, and remove the blocking warmup await (tasks FIFO-queue behind warmup, with safe disk-loading fallback).

## Benchmark results (latest commit: preload BPE data)

repomix repo root (999 files, 4-core, 10 runs each):

| Metric | Before | After | Improvement |
|--------|--------|-------|-------------|
| Median | 913ms | 855ms | **-58ms (-6.4%)** |
| p75 | 925ms | 864ms | **-61ms (-6.6%)** |

Pipeline shift (single profiled run):
- `calculateMetrics` start: 707ms → 522ms (**185ms earlier**)
- `pack()` end: 919ms → 802ms (**117ms faster**)

## Test plan

- [x] All 1128 tests passing
- [x] Lint clean (0 errors)
- [x] Reviewed by independent sub-agents for correctness, edge cases, and code quality
- [x] Verified: output correctness (999 files, correct token counts)
- [x] Verified: `tokenCountTree=true` still uses all workers and tokenizes all files
- [x] Verified: no regression on real secret detection
- [x] Verified: MCP server path unaffected
- [x] Verified: BPE preload failure degrades gracefully (workers load from disk)
- [x] Verified: backwards-compatible (optional parameter, fallback path intact)

https://claude.ai/code/session_01Cbzh8SssmdjmCsDJWaZ5tY

### CODE REVIEW

File: src/core/packager.ts [hunk 1]
```diff
@@ -69,6 +71,21 @@ export const pack = async (
...
+  //
+  // Since file count is unknown at this point, pass processConcurrency * TASKS_PER_THREAD
+  // to ensure getWorkerThreadCount allocates the maximum number of threads (one per core).
+  // This may over-allocate for very small repos, but the cost is bounded by processConcurrency
+  // threads and is negligible compared to the contention savings for typical repos.
+  const processConcurrency = deps.getProcessConcurrency();
+  const { taskRunner: metricsTaskRunner, warmupPromise: metricsWarmupPromise } = deps.createMetricsTaskRunner(
+    processConcurrency * 100,
+    config.tokenCount.encoding,
+  );
```
[the remark this answers is not in the document]
Comment (yamadashy) [reply src/core/packager.ts]: Fixed in fe2ecbe — moved the `try` block to start immediately after metrics worker pool creation, ensuring cleanup via `finally` covers the entire lifecycle including `searchFiles` and `sortPaths`.

---
_Generated by [Claude Code](https://claude.ai/code)_
